### PR TITLE
Enable running on GPU for oscar

### PR DESCRIPTION
**Pull request by hekaisheng**
## What do these changes do?

This PR aims to enable mars jobs running on GPUs.

### CODE REVIEW

File: mars/serialization/aio.py [hunk 1]
```diff
@@ -33,6 +35,23 @@ def __init__(self,
     def _get_buffers(self):
         headers, buffers = serialize(self._obj)
 
+        def _is_cuda_buffer(buf):  # pragma: no cover
+            try:
+                from cudf.core import Buffer as CPBuffer
```
Comment (qinxuye) [inline mars/serialization/aio.py:40]: Change to lazy_import.

File: mars/services/storage/handler.py [hunk 2]
```diff
@@ -120,16 +131,34 @@ async def batch_get(self, args_list, kwargs_list):
                 results.append(result)
         raise mo.Return(results)
 
+    def _get_default_level(self, obj):
+        obj = obj[0] if isinstance(obj, (list, tuple)) else obj
+        if self.highest_level != StorageLevel.GPU:
+            return self.highest_level
+        else:  # pragma: no cover
+            try:
+                import cudf, cupy
```
Comment (qinxuye) [inline mars/services/storage/handler.py:140]: Use lazy_import.

File: mars/services/storage/handler.py [hunk 3]
```diff
@@ -340,24 +371,26 @@ async def _fetch_remote(self,
             put_data_info_delays.append(
                 self._data_manager_ref.put_data_info.delay(
                     session_id, data_key, data_info, None))
-            fetch_tasks.append(self._clients[level].fetch(data_info.object_id))
+            fetch_tasks.append(self._clients[StorageLevel.REMOTE].fetch(data_info.object_id))
```
Comment (qinxuye) [inline mars/services/storage/handler.py:374]: Check coverage here.

File: mars/services/storage/handler.py [hunk 4]
```diff
@@ -448,18 +489,19 @@ async def spill(self,
...
+            await spill(size, level, self._band_name, self._data_manager_ref, self)
 
     async def list(self, level: StorageLevel) -> List:
-        return await self._data_manager_ref.list(level)
+        return await self._data_manager_ref.list(level, self._band_name)
 
     @mo.extensible
     async def unpin(self, session_id: str, data_key: str, error: str = 'raise'):
-        levels = await self._data_manager_ref.unpin(session_id, [data_key], error)
+        levels = await self._data_manager_ref.unpin(
```
Comment (qinxuye) [inline mars/services/storage/handler.py:503]: Why unpin not covered?

File: mars/services/task/analyzer/assigner.py [hunk 5]
```diff
@@ -52,6 +52,14 @@ def assign(self, cur_assigns: Dict[str, str] = None) -> Dict[ChunkData, BandType
             From node to band.
         """
 
+    def get_device_band_slots(self) -> Dict[BandType, int]:
+        if self._start_ops and all(op.gpu for op in self._start_ops):
```
Comment (qinxuye) [inline mars/services/task/analyzer/assigner.py:56]: No cover here.

File: mars/storage/cuda.py [hunk 6]
```diff
@@ -30,90 +32,197 @@
...
+    def _initialize_write(self):
+        self._had_write_headers = False
+        self._cur_buffer_index = 0
+        self._buffers = []
+        self._offset = 0
+
+    def read(self, size: int):
+        # we read cuda_header first and then read cuda buffers one by one,
+        # the return value's size is not exactly the specified size.
+        from cudf.core import Buffer
```
Comment (qinxuye) [inline mars/storage/cuda.py:107]: Use lazy import.

**Review by qinxuye — COMMENTED** [review]
I left some comments

File: mars/utils.py [hunk 7]
```diff
@@ -568,10 +571,18 @@ def sort_dataframe_result(df, result: pd.DataFrame) -> pd.DataFrame:
...
+                # cudf doesn't support inplace
+                result = result.sort_index()
         if hasattr(df, 'columns_value'):
             if getattr(df.columns_value, 'should_be_monotonic', False):
-                result.sort_index(axis=1, inplace=True)
+                try:
+                    result.sort_index(axis=1, inplace=True)
+                except TypeError:  # pragma: no cover
+                    # cudf doesn't support inplace
+                    result = result.sort_index(axis=1, inplace=True)
```
Comment (wjsi) [inline mars/utils.py:585]: Still `inplace=True`?

File: mars/services/storage/handler.py (hunk 2 above)
Comment (wjsi) [inline mars/services/storage/handler.py:140]: Use `lazy_import`.

**Review by wjsi — CHANGES_REQUESTED** [review]

**Review by qinxuye — APPROVED** [review]
LGTM

**Review by wjsi — APPROVED** [review]
LGTM

**Pull request merged**